### DISCUSSION TITLE
New metrics

### DISCUSSION THREAD

**Discussion by dubigrasu** [ChristopherKlay/StadiaEnhanced#89] [Ideas & Requests]
One of the problems I faced few times (and I saw it mentioned by many users) is the confusion regarding software or hardware decoding and the importance of proper video decoding and its impact on gaming experience.

Many times users are receiving the advice to enable or disable the hardware acceleration, and frequently they respond "how do I know that is enabled or even working?" or "which one is better?".
There are ways to check that of course, but they are rather cumbersome and not practical.

I think here Stadia Enhanced has the potential of solving that confusion with two new metrics: a) the type of decoding used (hardware or software), and b) the decoding times.
For example here I modified my local copy of the extension to display these two metrics, and you can see the impact of resolution, codecs, hardware/software decoding upon decoding times (I called it here "vdl", as in video decoding latency):
![stuff](https://user-images.githubusercontent.com/6303599/104129689-e00a9100-5375-11eb-9708-aafa423da3a0.png)
I think such information could greatly help the user in finding the best compromise/performance for a given system.

Not sure how exactly these two metrics should be titled, but after some time on my system I settled to this:
![Screenshot from 2021-01-10 20-44-00](https://user-images.githubusercontent.com/6303599/104132327-5498fc00-5385-11eb-8d27-d22516dd25d4.png)
Maybe they should be more explicit, or more verbose, or placed/formulated differently, but I'm sure an optimal arrangement can be found.

**Comment by ChristopherKlay** [comment]
Great idea!

Just to make sure, the "decoding time" in this case uses `totalDecodeTime/framesDecoded_in_ms`, similar to the statistics via `chrome://webrtc` correct?
I also never looked into hardware/software support values tbh, which one did you check for that?

**Reply by dubigrasu** [reply]
Yes, is `totalDecodeTime/framesDecoded_in_ms`.
For hw/sw is `decoderImplementation` in `chrome://webrtc`
But I think I'll just post some of the lines I modified, to be exact.

I have to warn you though, I'm no coder at all. I just stared at the code until it made some sense to me, googled some things and used the trial and error method until it worked. So what I did it may be a total crap, but anyway:

`var codec = tmp3.stat("googCodecName") + " " +  tmp3.stat("codecImplementationName").replace("FFmpeg", "software").replace("libvpx", "software").replace("ExternalDecoder", "hardware");`

`var dlatency = (tmp1.totalDecodeTime / tmp2.framesDecoded) * 1000;`

`enhanced_streamData = codec + ":" + " " + dlatency.toFixed(2) + "ms" + " &ensp;|&ensp;" + resolution + "&ensp;|&ensp;" + framesReceivedPerSecond.toFixed(1) + "fps&ensp;|&ensp;" + slatency + "ms  <span style='color: " + enhanced_connectionStatus + ";'>⬤</span>";`

I used "dlatency" to differentiate from "slatency", (again, I don't know how this works so I improvised).
Also, there may be other software codecs for  `decoderImplementation`, but I only know about FFmpeg and libvpx.

**Comment by ChristopherKlay** [comment]
No worries, everything is perfectly fine and this definitely helps the project (:

I did some research and the time is indeed called "Decoding Time", at least in most topics about drivers/webrtc i found.

The possible software codes shouldn't be an issue if we filter them out via;
```
if (tmp3.stat("codecImplementationName") == "ExternalDecoder") {
    var decodingType = "Hardware";
} else {
    var decodingType = "Software";
}
```
I already implemented the metric into the monitor, confirmed the results via chromes webrtc statistic and it will be part of `1.8.4` which will go live later today. (:

**Reply by dubigrasu** [reply]
That's great, many thanks!
